File: examples/java_project/src/main/java/com/example/pages/LoginPage.java
package com.example.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.PageFactory;
// VIOLATION: FORBIDDEN_IMPORT - test framework import in Page Object
import org.junit.jupiter.api.Assertions;

/**
 * Page Object with intentional gTAA violations for demonstration.
 *
 * Violations:
 * - FORBIDDEN_IMPORT: JUnit import in Page Object
 * - ASSERTION_IN_POM: Assertions inside Page Object methods
 */
public class LoginPage {

    private WebDriver driver;

    @FindBy(id = "username")
    private WebElement usernameField;

    @FindBy(id = "password")
    private WebElement passwordField;

    @FindBy(id = "loginBtn")
    private WebElement loginButton;

    @FindBy(css = ".error-message")
    private WebElement errorMessage;

    public LoginPage(WebDriver driver) {
        this.driver = driver;
        PageFactory.initElements(driver, this);
    }

    public void enterUsername(String username) {
        usernameField.clear();
        usernameField.sendKeys(username);
    }

    public void enterPassword(String password) {
        passwordField.clear();
        passwordField.sendKeys(password);
    }

    public void clickLogin() {
        loginButton.click();
    }

    public void login(String username, String password) {
        enterUsername(username);
        enterPassword(password);
        clickLogin();
    }

    // VIOLATION: ASSERTION_IN_POM
    public void verifyLoginSuccessful() {
        Assertions.assertTrue(driver.getCurrentUrl().contains("dashboard"));
    }

    // VIOLATION: ASSERTION_IN_POM
    public void verifyErrorMessage(String expectedMessage) {
        Assertions.assertEquals(expectedMessage, errorMessage.getText());
    }

    // VIOLATION: ASSERTION_IN_POM
    public void assertUserIsLoggedIn() {
        Assertions.assertNotNull(driver.findElement(By.id("userAvatar")));
    }
}

File: gtaa_validator/examples/csharp_project/Pages/LoginPage.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Support.PageObjects;
// VIOLATION: FORBIDDEN_IMPORT - test framework import in Page Object
using NUnit.Framework;

namespace ExampleProject.Pages
{
    /// <summary>
    /// Page Object with intentional gTAA violations for demonstration.
    ///
    /// Violations:
    /// - FORBIDDEN_IMPORT: NUnit import in Page Object
    /// - ASSERTION_IN_POM: Assert calls inside Page Object methods
    /// </summary>
    public class LoginPage
    {
        private readonly IWebDriver _driver;

        [FindsBy(How = How.Id, Using = "username")]
        private IWebElement UsernameField { get; set; }

        [FindsBy(How = How.Id, Using = "password")]
        private IWebElement PasswordField { get; set; }

        [FindsBy(How = How.Id, Using = "loginBtn")]
        private IWebElement LoginButton { get; set; }

        [FindsBy(How = How.CssSelector, Using = ".error-message")]
        private IWebElement ErrorMessage { get; set; }

        public LoginPage(IWebDriver driver)
        {
            _driver = driver;
            PageFactory.InitElements(driver, this);
        }

        public void EnterUsername(string username)
        {
            UsernameField.Clear();
            UsernameField.SendKeys(username);
        }

        public void EnterPassword(string password)
        {
            PasswordField.Clear();
            PasswordField.SendKeys(password);
        }

        public void ClickLogin()
        {
            LoginButton.Click();
        }

        public void Login(string username, string password)
        {
            EnterUsername(username);
            EnterPassword(password);
            ClickLogin();
        }

        // VIOLATION: ASSERTION_IN_POM
        public void VerifyLoginSuccessful()
        {
            Assert.That(_driver.Url, Does.Contain("dashboard"));
        }

        // VIOLATION: ASSERTION_IN_POM
        public void VerifyErrorMessage(string expectedMessage)
        {
            Assert.AreEqual(expectedMessage, ErrorMessage.Text);
        }

        // VIOLATION: ASSERTION_IN_POM
        public void AssertUserIsLoggedIn()
        {
            Assert.IsNotNull(_driver.FindElement(By.Id("userAvatar")));
        }
    }
}
